[PowerPoint] New articles for working with tables and shape bindings (#5105)
OfficeDev/office-js-docs-pr: docs/powerpoint/bind-shapes-in-presentation.md

---
title: Bind to shapes in a PowerPoint presentation
description: Learn how to bind shapes and access them from your add-in to keep them up-to-date.
ms.topic: how-to
ms.date: 03/31/2025
ms.localizationpriority: medium
---

# Bind to shapes in a PowerPoint presentation

Your PowerPoint add-in can bind to shapes to consistently access them through an identifier. Your add-in establishes a binding by calling `BindingCollection.add` and assigning a unique identifier. The identifier can be used at any time to reference the shape and access its properties. Creating bindings provides the following value to your add-in.

> - Establishes a relationship between the add-in and the shape in the document. Bindings are persisted in the document and can be accessed at a later time.

Reviewer: ```suggestion
- Establishes a relationship between the add-in and the shape in the document. Bindings are persistent in the document and can be accessed at a later time.
```
Persistent=adjective
Persisted=past tense verb

Reviewer: I'm just taking localization into consideration here. It still seems to me that persistent is the better word choice.

:::image type="content" source="../images/powerpoint-bind-shapes.png" alt-text="Diagram of binding to a star shape with the ID 'star' and binding to a pie chart with the ID 'pie'":::

## Scenario: Use bindings to sync with a data source

A common scenario for using bindings is to help users keep shapes up to date with a data source. Often when a user creates a presentation, they copy and paste images from the data source into the presentation. Over time, to keep the images up to date, they manually copy and paste the latest images from the data source. Your add-in can help automate this process by retrieving up-to-date images from the data source on the user’s behalf. When the user needs a shape fill updated, your add-in can use the binding to find the correct shape, and update the shape fill with the newer image.

In a general implementation, there are two components to consider for binding a shape in PowerPoint and updating it with a new image from a data source.

1. **The data source**. This is any source of data or asset library such as Microsoft SharePoint or Microsoft OneDrive.  

> 1. **The PowerPoint add-in**. The add-in gets data from the data source, based on what the user needs. It converts the data to a base64 formatted image. This is the only fill type the bound shape can accept. It inserts a shape upon the user’s request and binds it with a unique identifier. Then it fills the shape with the base64 image based on the original data source. Shapes are updated upon the user’s request and the add-in uses the binding identifier to find the shape and update the image with the last saved base64 image.

Reviewer: ```suggestion
1. **The PowerPoint add-in**. The add-in gets data from the data source based on what the user needs. It converts the data to a base64 formatted image. This is the only fill type the bound shape can accept. It inserts a shape upon the user’s request and binds it with a unique identifier. Then it fills the shape with the base64 image based on the original data source. Shapes are updated upon the user’s request and the add-in uses the binding identifier to find the shape and update the image with the last saved base64 image.
```

```javascript
async function updateBinding(bindingId, image) {
    await PowerPoint.run(async (context) => {
        try {
            // Get the shape based on binding ID. 
            const myShape = context.presentation.bindings
                .getItem(bindingId)
                .getShape();

            // Update the shape to latest image. 
            myShape.fill.setImage(image);
            await context.sync();

        } catch (err) {
            console.error(err);
        }
    });
} 
```

## Delete a binding

The following sample shows how to delete a binding by deleting it from the bindings collection.

```javascript
async function deleteBinding(bindingId) {
    await PowerPoint.run(async (context) => {
        context.presentation.bindings.getItemAt(bindingId).delete();
        await context.sync();
    });
}
```

## Load bindings

> When a presentation is opened and your add-in first loads, you can load all the bindings to continue working with them. The following code shows how to load all bindings in a presentation and display them in the console.

Reviewer: ```suggestion
When you open presentation and your add-in first loads, you can load all the bindings to continue working with them. The following code shows how to load all bindings in a presentation and display them in the console.
```
